database/abusescanner.go

```diff
@@ -0,0 +1,429 @@
+	AbuseEmail struct {
+		ID        primitive.ObjectID `bson:"_id"`
+		UID       string             `bson:"email_uid"`
+		UIDRaw    uint32             `bson:"email_uid_raw"`
+		Body      []byte             `bson:"email_body"`
+		From      string             `bson:"email_from"`
+		Subject   string             `bson:"email_subject"`
+		MessageID string             `bson:"email_message_id"`
+
+		// fields set by parser
+		ParsedAt    time.Time   `bson:"parsed_at"`
+		ParseResult AbuseReport `bson:"parseResult"`
+		Parsed      bool        `bson:"parsed"`
+
+		// fields set by blocker
+		BlockedAt   time.Time `bson:"blocked_at"`
+		BlockResult []string  `bson:"blockResult"`
+		Blocked     bool
+
+		InsertedAt time.Time `bson:"inserted_at"`
+		Finalized  bool
+	}
+
+	// AbuseReport contains all information about an abuse report.
+	AbuseReport struct {
+		Skylinks []string
+		Reporter AbuseReporter
+		Sponsor  string
+		Tags     []string
+	}
+
+	// AbuseReporter encapsulates some information about the reporter.
+	AbuseReporter struct {
+		Name         string
+		Email        string
+		OtherContact string
+	}
+
+	// abuseEmailLock represents a lock on an abuse email.
+	abuseEmailLock struct {
+		staticClient         *lock.Client
+		staticEmailUID       string
+		staticPortalHostname string
+	}
+)
+
+// String returns a string representation of the abuse email
+func (a AbuseEmail) String() string {
+	var sb strings.Builder
+	pr := a.ParseResult
+	sb.WriteString("\nAbuse Scanner Report:\n")
+
+	sb.WriteString("\nReporter:\n")
+	sb.WriteString(fmt.Sprintf("Name: %v\n", pr.Reporter.Name))
+	sb.WriteString(fmt.Sprintf("Email: %v\n", pr.Reporter.Email))
+
+	sb.WriteString("\nTags:\n")
+	for _, tag := range pr.Tags {
+		sb.WriteString(tag + "\n")
+	}
+
+	allBlocked := true
+	sb.WriteString("\nSkylinks:\n")
+	for i, skylink := range pr.Skylinks {
+		var parts []string
+		if a.BlockResult[i] == AbuseStatusBlocked {
+			parts = []string{AbuseStatusBlocked, skylink}
+		} else {
+			parts = []string{AbuseStatusNotBlocked, skylink, a.BlockResult[i]}
+			allBlocked = false
+		}
+		switch a.BlockResult[i] {
+		case AbuseStatusBlocked:
+		default:
+		}
+		sb.WriteString(fmt.Sprintf("%s\n", strings.Join(parts, " | ")))
+	}
+
+	sb.WriteString("\nSummary:\n")
+	if len(pr.Skylinks) == 0 {
+		sb.WriteString("FAILURE - no skylinks found.\n")
+	} else {
+		if allBlocked {
+			sb.WriteString("SUCCESS\n")
+		} else {
+			sb.WriteString("FAILURE - not all skylinks blocked.\n")
+		}
+	}
+
+	return sb.String()
+}
+
+// NewAbuseScannerDB returns an instance of the Mongo DB.
+func NewAbuseScannerDB(connectionString, portalHostName string, logger *logrus.Logger) (*AbuseScannerDB, error) {
+	// create the client
+	opts := options.Client().ApplyURI(connectionString)
+	client, err := mongo.NewClient(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	// create a context with timeout
+	ctx, cancel := context.WithTimeout(context.Background(), mongoDefaultTimeout)
```

Comment:
Shouldn't we use the global context that we create in `main.go` and add on top of that, so once that's cancelled all others avalanche as a result?